src/index.ts

```diff
@@ -1,19 +1,37 @@
-import {DependencyList, useEffect, useState} from 'react'
+import { DependencyList, useRef, useEffect, useState, useReducer, useMemo } from 'react'
+
+export function useAsyncMemo<T>(factory: () => Promise<T>, deps?: DependencyList, initial?: T) {
+  let [val, setVal] = useState(initial)
 
-export function useAsyncMemo<T>(factory: () => Promise<T> | undefined | null, deps: DependencyList, initial: T = undefined): T {
-  const [val, setVal] = useState<T>(initial)
   useEffect(() => {
-    let cancel = false
-    const promise = factory()
-    if (promise === undefined || promise === null) return
-    promise.then((val) => {
-      if (!cancel) {
-        setVal(val)
+    let pending = true
+    factory().then(res => pending && setVal(res))
+
+    return () => {
+      pending = false
+    }
+  }, deps)
+
+  return val
+}
+
+export function useAsync<T>(factory: () => Promise<T>, deps?: DependencyList, initial?: T) {
```

Comment:
What's the difference between this `useAsync` hook and `useAsyncMemo`?

Reply:
Seems they work same for users while the implementation is different?

Reply:
`useAsync` does not save the previous result

Reply:
Oh, I got it. When `deps` changes, `val.current` will get reseted to `initial`.
But I think this package should focus on `useAsyncMemo` and `useAsync` should be shipped as a separate package.
What's more, I'm current working on another Hook `useAsyncEffect`. The logic of `useAsync` can be implemented like this:

```js
const [val, setVal] = useState()
useAsyncEffect(async function*() {
  setVal(initial)
  const nextVal = await fetch()
  yield
  setVal(nextVal)
}, [x, xx])
```

Reply:
Here is the PR of `useAsyncEffect`: https://github.com/alibaba/hooks/pull/958

Reply:
This solution will create unwanted extra rerenders, that's why I use useRef
I also think that you should not put useAsync in a separate package (naming problems)